Source/FieldSolver/WarpXPushFieldsHybridPIC.cpp

```diff
@@ -0,0 +1,254 @@
+                0.5_rt, *current_fp[lev][idim], 0,
+                0, 1, current_fp_temp[lev][idim]->nGrowVect()
+            );
+        }
+    }
+
+    // Calculate the electron pressure at t=n using rho^n
+    CalculateElectronPressure(DtType::FirstHalf);
+
+    // Push the B field from t=n to t=n+1/2 using the current and density
+    // at t=n, while updating the E field along with B using the electron
+    // momentum equation
+    for (int sub_step = 0; sub_step < sub_steps; sub_step++)
+    {
+        CalculateCurrentAmpere();
+        HybridPICSolveE(DtType::FirstHalf);
+        EvolveB(0.5 / sub_steps * dt[0], DtType::FirstHalf);
+        FillBoundaryB(guard_cells.ng_FieldSolver, WarpX::sync_nodal_points);
+    }
+
+    // Average rho^{n} and rho^{n+1} to get rho^{n+1/2} in rho_fp_temp
+    for (int lev = 0; lev <= finest_level; ++lev)
+    {
+        MultiFab::LinComb(
+            *rho_fp_temp[lev], 0.5_rt, *rho_fp_temp[lev], 0,
+            0.5_rt, *rho_fp[lev], 0, 0, 1, rho_fp_temp[lev]->nGrowVect()
+        );
+    }
+
+    // Calculate the electron pressure at t=n+1/2
+    CalculateElectronPressure(DtType::SecondHalf);
+
+    // Now push the B field from t=n+1/2 to t=n+1 using the n+1/2 quantities
+    for (int sub_step = 0; sub_step < sub_steps; sub_step++)
+    {
+        CalculateCurrentAmpere();
+        HybridPICSolveE(DtType::SecondHalf);
+        EvolveB(0.5 / sub_steps * dt[0], DtType::SecondHalf);
+        FillBoundaryB(guard_cells.ng_FieldSolver, WarpX::sync_nodal_points);
+    }
+
+    // Extrapolate the ion current density to t=n+1 using
+    // J_i^{n+1} = 1/2 * J_i^{n-1/2} + 3/2 * J_i^{n+1/2}, and recalling that
+    // now current_fp_temp = J_i^{n} = 1/2 * (J_i^{n-1/2} + J_i^{n+1/2})
+    for (int lev = 0; lev <= finest_level; ++lev)
+    {
+        for (int idim = 0; idim < 3; ++idim) {
+            MultiFab::LinComb(
+                *current_fp_temp[lev][idim],
+                -1._rt, *current_fp_temp[lev][idim], 0,
+                2._rt, *current_fp[lev][idim], 0,
+                0, 1, current_fp_temp[lev][idim]->nGrowVect()
+            );
+        }
+    }
+
+    // Calculate the electron pressure at t=n+1
+    CalculateElectronPressure(DtType::Full);
+
+    // Update the E field to t=n+1 using the extrapolated J_i^n+1 value
+    CalculateCurrentAmpere();
+    HybridPICSolveE(DtType::Full);
+
+    // Copy the rho^{n+1} values to rho_fp_temp and the J_i^{n+1/2} values to
+    // current_fp_temp since at the next step those values will be needed as
+    // rho^{n} and J_i^{n-1/2}.
+    for (int lev = 0; lev <= finest_level; ++lev)
+    {
+        MultiFab::Copy(*rho_fp_temp[lev], *rho_fp[lev],
+                        0, 0, 1, rho_fp_temp[lev]->nGrowVect());
+        for (int idim = 0; idim < 3; ++idim) {
+            MultiFab::Copy(*current_fp_temp[lev][idim], *current_fp[lev][idim],
+                           0, 0, 1, current_fp_temp[lev][idim]->nGrowVect());
```

Comment:
I always find it useful to add a brief inline comment explaining what the various "0" and "1" represent in these AMReX function calls (it would be nice if C++ had a Python-like syntax such as `f(arg1=val1, arg2=val2, ...)`). I know we don't do it consistently throughout the code, but if you too think that it's a good idea, this could be a good place to start. If you do so, it could be good to do it consistently at least in this file.

Reply:
Thanks for continuing the review!! I agree that such comments would be helpful for the next person trying to read the code. I'll add the comments (also for the next point) and re-ping you for a review when it is done.